x-pack/otel/exporter/logstashexporter/internal/event.go

```diff
@@ -0,0 +1,94 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"go.opentelemetry.io/collector/client"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+)
+
+func parseEvent(ctx context.Context, logRecord *plog.LogRecord) (beat.Event, error) {
+	metadata := parseEventMetadata(ctx)
+	if !isBeatsEvent(metadata) {
+		return beat.Event{}, consumererror.NewPermanent(errors.New("invalid beats event metadata"))
+	}
+
+	fields, ok := parseEventFields(logRecord)
+	if !ok {
+		return beat.Event{}, consumererror.NewPermanent(errors.New("invalid beats event body, expected a map, got: " + logRecord.Body().Type().String()))
+	}
+
+	timestamp, ok := parseEventTimestamp(fields)
+	if !ok {
+		timestamp = logRecord.ObservedTimestamp().AsTime()
+	}
+
+	return beat.Event{
+		Timestamp: timestamp,
+		Meta:      metadata,
+		Fields:    fields,
+	}, nil
+}
+
+func parseEventFields(logRecord *plog.LogRecord) (map[string]any, bool) {
+	if logRecord.Body().Type() != pcommon.ValueTypeMap {
+		return nil, false
+	}
+	return logRecord.Body().Map().AsRaw(), true
+}
```

Comment:
You can probably use `otelmap.ToMapstr` here. We can probably have a variant in `otelmap` that returns a `beat.Event` then you do not need most of the parse logic here, and it stays self-contained.